overlord/ifacestate/ifacestate.go

```diff
@@ -35,17 +36,90 @@ import (
 	"github.com/snapcore/snapd/snap"
 )
 
-var noConflictOnConnectTasks = func(task *state.Task) bool {
-	return task.Kind() != "connect" && task.Kind() != "disconnect"
+func noConflictOnConnectTasks(chg *state.Change) func(task *state.Task) bool {
+	checkConflict := func(task *state.Task) bool {
+		if chg != nil && task.Change() != nil {
+			// if same change, then return false (no conflict)
+			return chg.ID() != task.Change().ID()
+		}
+		return task.Kind() != "connect" && task.Kind() != "disconnect"
+	}
+	return checkConflict
+}
+
+var connectRetryTimeout = time.Second * 5
+
+func checkConnectConflicts(st *state.State, change *state.Change, plugSnap, slotSnap string) error {
```

Comment:
thanks for doing it this way, seems the right direction, I think as I discuss below that we need to distinguish the auto-connect vs not case logic a bit more, we probably need to pass in the being install snap name in the auto-connect case